src/core/dataset_container.jl

```diff
@@ -0,0 +1,179 @@
+struct DatasetContainer{T}
+    duals::Dict{ConstraintKey, T}
+    aux_variables::Dict{AuxVarKey, T}
+    variables::Dict{VariableKey, T}
+    parameters::Dict{ParameterKey, T}
+    expressions::Dict{ExpressionKey, T}
+end
+
+function DatasetContainer{T}() where {T <: AbstractDataset}
+    return DatasetContainer(
+        Dict{ConstraintKey, T}(),
+        Dict{AuxVarKey, T}(),
+        Dict{VariableKey, T}(),
+        Dict{ParameterKey, T}(),
+        Dict{ExpressionKey, T}(),
+    )
+end
+
+function Base.empty!(container::DatasetContainer)
+    for field in fieldnames(DatasetContainer)
+        field_dict = getfield(container, field)
+        for val in values(field_dict)
+            empty!(val)
+        end
+    end
+    @debug "Emptied the store" _group = LOG_GROUP_SIMULATION_STORE
+    return
+end
+
+function get_duals_values(container::DatasetContainer)
+    return container.duals
+end
+
+function get_aux_variables_values(container::DatasetContainer)
+    return container.aux_variables
+end
+
+function get_variables_values(container::DatasetContainer)
+    return container.variables
+end
+
+function get_parameters_values(container::DatasetContainer)
+    return container.parameters
+end
+
+function get_expression_values(container::DatasetContainer)
+    return container.expressions
+end
+
+function get_dataset_keys(container::DatasetContainer)
+    return Iterators.flatten(
+        keys(getfield(container, f)) for f in fieldnames(DatasetContainer)
+    )
+end
+
+function get_dataset(container::DatasetContainer, key::OptimizationContainerKey)
+    datasets = getfield(container, get_store_container_type(key))
+    return datasets[key]
+end
+
+function set_dataset!(
+    container::DatasetContainer{T},
+    key::OptimizationContainerKey,
+    val::T,
```

Comment:
The `T` doesn't look important. Can you remove it?
```
val::AbstractDataset
```

Reply:
it does, as it forces that the parametric type of the container matches the type of value being written

Reply:
That is technically not necessary because the addition to `datasets` will fail if the type is wrong. But I guess I don't have a problem with it.